### PR TITLE
In-place transforms

### PR DESCRIPTION
**Pull request by CFlaniganMide**
I'm breaking up the faster array stuff into a few parts.

This branch adds this functionality:
- `Transform._fastCoeffs`, which are the coefficients of a polynomial with substitutions after it's been expanded.
- `Transform.inplace`, calculates the values (but not timestamps) for a transform.  If the keyword argument `out` is specified, the values are inserted into that array rather than generating a new array.  Where possible, numpy in-place math is used to reduce memory and computational overhead.  Utilizes `_fastCoeffs` to reduce the amount of arithmetic.
- `EventArray.getMean` calculates the mean across a certain time range.  Caches the result for the full file.

It modifies the following functionality:
- `Transform` calls will use the full mean of secondary data source by default instead of per-block.
- `EventArray._makeBlockEvents` uses the inplace function.

Based on the current implementation, `Univariate` and `Bivariate` only support constant and linear equations.  Higher order polynomials are definitely doable, but are more complex and currently aren't used.

TODOs:
- Finish implementing a few corner cases
- Tests

### CODE REVIEW

**Comment by CFlaniganMide** [comment]
Had a conversation with David, here's the broad plan to move forward with the API:
- Individual `Bivariate` transforms will keep track of weather or not to use the full or mean of their secondary channel with a `useMean` attribute
  - `Univariate` channel will generate a warning for the user if `useMean` is assigned a value
- When a `Bivariate` polynomial updates it's `useMean` value, it will update it's parents appropriately.
- `CombinedPoly` and `PolyPoly` will add themselves to their childrens' `_parents` attribute, which should be a `weakref.WeakSet`

**Comment by CFlaniganMide** [comment]
addendum: `_parents` could cause some confusion, something like `_referencers` or `_watchers` would avoid this.

**Review by StokesMIDE — COMMENTED** [review]
In the tests, it would be useful to have some comments describing what's being tested. All the tests should get this, but we might as well start with what's new in this PR.

**Comment by CFlaniganMide** [comment]
> 
> 
> In the tests, it would be useful to have some comments describing what's being tested. All the tests should get this, but we might as well start with what's new in this PR.

Yeah, agreed.  I'm working on the unit tests for the new stuff right now, so I'll add docs for the new stuff.

File: idelib/transforms.py [hunk 1]
```diff
@@ -219,18 +270,29 @@ def __init__(self, coeffs, calId=None, dataset=None, reference=0,
...
         self._timeOffset = 0
+        self._watchers = weakref.WeakSet()
         
         self.attributes = attributes
         
         self._build()
+
+
+    def __hash__(self):
+        return hash((self._coeffs, self._references))
```
Comment (StokesMIDE) [inline idelib/transforms.py:295]: The coefficients and references won't be unique, even in the same file... should you at least check the ID as well?

Comment (CFlaniganMide) [reply idelib/transforms.py:295]: I've based the hash function(s) on the recommendation of the recommendation of the [Python docs for overriding the __hash__ function](https://docs.python.org/3/reference/datamodel.html#object.__hash__).  Basically, the recommendation is that the hash function combines the things that are compared in the `__eq__` method to generate a new hash, so that any objects that are equal to each other have the same hash.

Comment (StokesMIDE) [reply idelib/transforms.py:295]: I thought the ID was part of the `__eq__()` comparison, but that doesn't appear to be the case. It really seems like should be.

Comment (CFlaniganMide) [reply idelib/transforms.py:295]: That would definitely make sense.  Do you want me to add both those changes?

Comment (StokesMIDE) [reply idelib/transforms.py:295]: Yes, I think it is a good addition, although I don't know what effect it will have on the tests. I think this `__eq__()` was added when other classes got them for the sake of testing.

**Comment by CFlaniganMide** [comment]
After some testing on slamsticklab's development branch, the only thing to break was the pressure-altitude conversion.  @StokesMIDE should this be treated as blocking, or should it be an issue to deal with later?

**Comment by StokesMIDE** [comment]
> After some testing on slamsticklab's development branch, the only thing to break was the pressure-altitude conversion. @StokesMIDE should this be treated as blocking, or should it be an issue to deal with later?

Personally, I think it's fairly important; I'd consider it a blocking issue. It's less of an issue for the library itself, but many Lab users use the feature.

**Comment by CFlaniganMide** [comment]
> 
> 
> > After some testing on slamsticklab's development branch, the only thing to break was the pressure-altitude conversion. @StokesMIDE should this be treated as blocking, or should it be an issue to deal with later?
> 
> Personally, I think it's fairly important; I'd consider it a blocking issue. It's less of an issue for the library itself, but many Lab users use the feature.

I definitely agree it's important.  Since the bug was in idelib code (specifically how `CombinedPoly` didn't have a case for anything other than linear transforms) I made the fixes here.

**Comment by CFlaniganMide** [comment]
@StokesMIDE @CrepeGoat I can't actually re-request reviews, so I'm just pinging.  I'd like to get this through today if possible.

File: testing/test_transforms.py [hunk 2]
```diff
@@ -0,0 +1,248 @@
...
+                    5,
+                    dtype=np.uint64,
+                    )
+            y = np.fromiter(
+                    (eventList[eventList.getEventIndexNear(t)][1] for t in timestamps),
+                    dtype=np.float64,
+                    )
+        else:
+            timestamps = None
+            y = eventList.getMean()
```
Comment (CrepeGoat) [inline testing/test_transforms.py:201]: so `useTimestamp=False, useMean=False` sets `y = eventList.getMean()`? From my alien eyes I'd assume `useMean=False` would make it so `getMean` doesn't get called; can you just confirm for me that this is intended?

**Review by CrepeGoat — COMMENTED** [review]
overall seems fine. a lot of the math was hard to focus on w/o zoning out, so I'm kind of trusting that the tests check the right things and that you got it right the first time... :\

I did have a handful of questions; I'd appreciate responses there but otherwise I'm fine giving the OK.

also I made two brevity changes; feel free to revert if you feel strongly about them, but I think that's easier than telling you to make them yourself.

File: idelib/transforms.py [hunk 3]
```diff
@@ -22,6 +27,11 @@
 #===============================================================================
 
 
+def _binomial(n, k):
+    """ Binomial coefficient of n choose k"""
+    return math.factorial(n)/(math.factorial(k)*math.factorial(n-k))
+
```
[the remark this answers is not in the document]
Comment (CFlaniganMide) [reply idelib/transforms.py:33]: I implemented this function a while ago, I'm going to delete it.  Turns out numpy has polynomial math implemented anyway so it's redundant.  That being said, I can imagine someone deciding to use a taylor series at some point, so arbitrary order polynomials would be nice eventually.

File: idelib/transforms.py [hunk 4]
```diff
@@ -108,16 +120,113 @@ def isValid(self, session=None, noBivariates=False):
...
+            values = float(values)
+        elif out is None:
+            out = np.zeros_like(values, dtype=np.float64)
+
+        if len(self._variables) == 1:
+
+            if scalar:
+                out = self.function(values)
+            else:
+                out[:] = self.function(values)
```
[the remark this answers is not in the document]
Comment (CFlaniganMide) [reply idelib/transforms.py:191]: I might change things to be more aware of 0d arrays, but right now they're definitely an edge case.  This is for a situation where the value passed in is just a regular int/float value, and the user is expecting the same to be returned.

File: testing/test_transforms.py [hunk 5]
```diff
@@ -0,0 +1,248 @@
...
+    tf = Univariate(coeffs, reference=reference)
+    return tf
+
+
+@pytest.fixture()
+def ssx66115():
+    testDataset = importer.openFile(makeStreamLike('./testing/SSX66115.IDE'))
+    importer.readData(testDataset)
+    yield testDataset
+    testDataset.close()
```
[the remark this answers is not in the document]
Comment (CFlaniganMide) [reply testing/test_transforms.py:50]: The function `makeStreamLike` caches file data in memory to reduce test execution time.  As for not using `importFile`, that's mostly because `makeStreamLike` returns a stream-like object, but `importFile` call's open on the file name passed in, and it was easier to just circumvent that call.

File: idelib/dataset.py [hunk 6]
```diff
@@ -2500,6 +2502,34 @@ def getValueAt(self, at, outOfRange=False, display=False):
...
+        if not self.hasMinMeanMax:
+            self._computeMinMeanMax()
+
+        if startTime is None and endTime is None:
+            if self._mean is not None:
+                return self._mean
+
+        means = self.arrayMinMeanMax(startTime, endTime, times=False,
+                                     display=display, iterator=iterator)[1]
+        mean = np.sum([d.numSamples*mean for d, mean in zip(self._data, means[0])])/np.sum(d.numSamples for d in self._data)
```
[the remark this answers is not in the document]
Comment (CFlaniganMide) [reply idelib/dataset.py:2527]: actually, I just realized that since this is a new method, I probably just copied the docstring from `getmin` and forgot to update it.

File: idelib/transforms.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (CrepeGoat) [reply idelib/transforms.py:191]: My comment was less "we should support 0-D arrays" and more "why not represent scalars as 0-D arrays?" They should behave the same way in any numpy operations, and it'd cut down on special-case logic.

File: idelib/dataset.py [hunk 7]
```diff
@@ -2500,6 +2502,34 @@ def getValueAt(self, at, outOfRange=False, display=False):
...
+            :return: The event with the minimum value.
+        """
+        if not self.hasMinMeanMax:
+            self._computeMinMeanMax()
+
+        if startTime is None and endTime is None:
+            if self._mean is not None:
+                return self._mean
+
+        means = self.arrayMinMeanMax(startTime, endTime, times=False,
```
[the remark this answers is not in the document]
Comment (CrepeGoat) [reply idelib/dataset.py:2525]: ...unless you pass `times=False`, then they don't... thanks for mentioning that 👍

File: idelib/dataset.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (CrepeGoat) [reply idelib/dataset.py:2527]: oh I didn't see the `times=False` bit. so that would change my most recent snippet to use `means` instead of `means[1:]`, making it look like this?
```python
        means = self.arrayMinMeanMax(startTime, endTime, times=False,
                                     display=display, iterator=iterator)[1]
        mean = np.amin(np.average(means, weights=[d.numSamples for d in self._data], axis=-1))
```

File: testing/test_transforms.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (CrepeGoat) [reply testing/test_transforms.py:50]: > The function makeStreamLike caches file data in memory to reduce test execution time.

Ah, yep that seems good

> As for not using importFile, that's mostly because ... `importFile` calls open on the file name passed in

...ew. yeah it'd be nice to change that. maybe during development on the "universal file opener" function?

[the remark this answers is not in the document]
Comment (CrepeGoat) [reply testing/test_transforms.py:50]: ...actually now that I look at it, you can set a pytest fixture's yield/return value to be shared across multiple tests (as opposed to being called and re-generated for every individual test) [by setting the fixture's scope parameter](https://docs.pytest.org/en/6.2.x/fixture.html#scope-sharing-fixtures-across-classes-modules-packages-or-session). 

not critical, but it looks like a more idiomatic way to cache results for test fixtures. one of these days we could maybe get around to using that 🤷

File: testing/test_transforms.py (hunk 2 above)
Comment (CrepeGoat) [reply testing/test_transforms.py:201]: I also forgot to mention that I swapped this here for `y = 0` and the tests still passed. so maybe this isn't super critical?

File: idelib/transforms.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (CFlaniganMide) [reply idelib/transforms.py:191]: Just did some quick testing, since that's an interesting idea.  Unfortunately, the following code is invalid:

```python
a = np.array(1)
a[:] = 2
```

Basically, you can't index into a 0-D array.

I'm going to play around with this a bit more and see if I can make it work a different way, since it looks like type promotion will always turn it back into an int/float.

File: idelib/dataset.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (CFlaniganMide) [reply idelib/dataset.py:2527]: yeah, that looks good, I just made a change to the code and I'll be pushing it.  Only difference is that the docstring was wrong, for Channels it's the mean of all the values in the channel, not the min of the means of the Subchannels.

File: testing/test_transforms.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (CFlaniganMide) [reply testing/test_transforms.py:50]: Yeah, I'll take a look, I think there's a better way to cache.  That being said, I'd point out a few issues I've found with fixtures, some of which are very specific to this case:
- The value that gets re-used is a reference, so it needs to be recalculated every time, rather than just re-passed.
- When a test uses a fixture, it also needs to be able to see every other fixture referenced by that fixture and so on.  If you have too many fixtures, then you end up with enormous blocks of imports, or `from x import *` which I don't like.  
- Adding narrow helper functions to the scope of what's "correct" in Pytest allows you to make functions to parameterize fixtures or tests, since fixtures can't be called in pytest marks.

File: testing/test_transforms.py (hunk 2 above)
Comment (CFlaniganMide) [reply testing/test_transforms.py:201]: I've interpreted `useMean` as "use the mean instead of the timestamps", so `useMean=False` is not the same as `noBivariates=True`.

File: testing/test_dataset.py [hunk 8]
```diff
@@ -1196,24 +1196,22 @@ def testGetItem(self):
...
+        # # and return None
+        # eventArray.configure_mock(
+        #     _fullXform=Univariate((None, 0)),
+        #     _getBlockRollingMean=lambda blockIdx: None,
+        #     hasSubchannels=True,
+        # )
+        # self.assertEqual(EventArray.__getitem__(eventArray, 0), None)
+        # self.assertEqual(EventArray.__getitem__(eventArray, 1), None)
+        # self.assertEqual(EventArray.__getitem__(eventArray, 2), None)
+        # self.assertEqual(EventArray.__getitem__(eventArray, 3), None)
```
[the remark this answers is not in the document]
Comment (CFlaniganMide) [reply testing/test_dataset.py:1209]: No, these tests pass `None` as a coefficient into a `Univariate` constructor call.  With the changes made to the transforms, `None` is no longer a valid coefficient.

**Comment by CFlaniganMide** [comment]
> 
> 
> so a lot of the math was hard to focus on w/o zoning out, so I'm kind of trusting that the tests check the right things and that you got it right the first time... :\
> 
> aside from that, I do have a handful of minor questions. I'd appreciate responses there, but otherwise I'm fine giving the OK.
> 
> also I made two brevity changes; feel free to revert if you feel strongly about them, but I think that's easier than telling you to make them yourself.

Did my best to respond to everything, also reverted one of the changes you made to `test_transforms.py`, since I think it makes it harder to read.  That being said, David and I spent some time talking about pytest parameterization and clarity, and that's definitely a larger discussion.

File: idelib/transforms.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (CFlaniganMide) [reply idelib/transforms.py:186]: This looks like a cleaner solution, but especially given the age of this PR I'd prefer to document this in an issue and add it in later.

File: testing/test_transforms.py (hunk 2 above)
Comment (CrepeGoat) [reply testing/test_transforms.py:209]: as long as it's intentional that's fine

File: testing/test_transforms.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (CrepeGoat) [reply testing/test_transforms.py:49]: yeah I dug into it a bit; it's too complicated with `unittest.TestCase` classes. all good 👍

File: idelib/transforms.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (CrepeGoat) [reply idelib/transforms.py:186]: yeah `a[:] = 2` will fail but `a[...] = 2` should work fine.

But I agree this should go in an issue, and not hold up this PR.

File: idelib/dataset.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (CFlaniganMide) [reply idelib/dataset.py]: Ok, double checked, I think something got lost translating this to the new method.  I don't 100% know why, but the math checks out now.

[the remark this answers is not in the document]
Comment (CrepeGoat) [reply idelib/dataset.py]: sweet! looks good now
